src/ecc.rs

```diff
@@ -0,0 +1,392 @@
+// Copyright 2023 - 2024 Simo Sorce, Jakub Jelen
+// See LICENSE.txt file for terms
+
+use super::attribute;
+use super::error;
+use super::interface;
+use super::object;
+use super::{attr_element, bytes_attr_not_empty, err_rv};
+
+use attribute::{from_bool, from_bytes};
+use error::{KError, KResult};
+use interface::*;
+use object::{
+    CommonKeyFactory, OAFlags, Object, ObjectAttr, ObjectFactories,
+    ObjectFactory, ObjectType, PrivKeyFactory, PubKeyFactory,
+};
+
+use once_cell::sync::Lazy;
+use std::fmt::Debug;
+
+pub const MIN_EC_SIZE_BITS: usize = 256;
+pub const MAX_EC_SIZE_BITS: usize = 521;
+
+#[derive(Debug)]
+pub struct ECCPubFactory {
+    attributes: Vec<ObjectAttr>,
+}
+
+impl ECCPubFactory {
+    pub fn new() -> ECCPubFactory {
+        let mut data: ECCPubFactory = ECCPubFactory {
+            attributes: Vec::new(),
+        };
+        data.attributes.append(&mut data.init_common_object_attrs());
+        data.attributes
+            .append(&mut data.init_common_storage_attrs());
+        data.attributes.append(&mut data.init_common_key_attrs());
+        data.attributes
+            .append(&mut data.init_common_public_key_attrs());
+        data.attributes.push(attr_element!(CKA_EC_PARAMS; OAFlags::AlwaysRequired | OAFlags::Unchangeable; from_bytes; val Vec::new()));
+        data.attributes.push(attr_element!(CKA_EC_POINT; OAFlags::RequiredOnCreate | OAFlags::UnsettableOnGenerate | OAFlags::Unchangeable; from_bytes; val Vec::new()));
+        data
+    }
+}
+
+impl ObjectFactory for ECCPubFactory {
+    fn create(&self, template: &[CK_ATTRIBUTE]) -> KResult<Object> {
+        let obj = self.default_object_create(template)?;
+
+        bytes_attr_not_empty!(obj; CKA_EC_PARAMS);
+        bytes_attr_not_empty!(obj; CKA_EC_POINT);
```

Comment:
Either of these would require parsing the ec params and also the ec point using the asn1 library, which is now done in the `src/ossl/ecc.c`. I think it can be moved, but OpenSSL does this anyway when the key is imported. The only advantage might be that the debugging might  be easier for bad keys, but we would have to reimplement quite some stuff from openssl.
Checking the length of point based on the parameters might make sense though.

Reply:
If the checks are already made elsewhere I am fine, but I am concerned about importing a key that will never work later.
That said I also do very little checking for RSA, so we can probably just defer better checking to later.

I do not have an issue with generate because that one is self checked in openssl, but create is just "import", unfortunately PKCS#11 calls import "create" which can be misleading.